### PR TITLE
Issue/jperf 535

### PR DESCRIPTION
**Pull request by dagguh**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -23,6 +23,14 @@ Dropping a requirement of a major version of a dependency is a new contract.
 ## [Unreleased]
 [Unreleased]: https://github.com/atlassian/virtual-users/compare/release-3.9.1...master
 
+### Added
+- Add `TimeControllingUserGenerator` for synchronizing user generators across threads and nodes.
```
Comment (dagguh) [inline CHANGELOG.md]: TBH it's fully implementable by users, so we don't have to provide it right now. I can cut it out of the PR.

File: src/main/kotlin/com/atlassian/performance/tools/virtualusers/LoadTest.kt [hunk 2]
```diff
@@ -58,18 +56,31 @@ internal class LoadTest(
...
             nodeCounter.dump(it)
         }
         logger.debug("Dumped node's counts to $nodesDump")
     }
 
+    private fun generateUsers(): List<User> {
+        val pool = Executors.newCachedThreadPool { runnable ->
+            Thread(runnable, "user-generation-${runnable.hashCode()}")
+        }
+        val deadline = now() + behavior.maxOverhead
```
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/virtualusers/LoadTest.kt]: Naturally, `maxOverhead` is not entirely dedicated to user generation, but it's still a much better timeout than infinity or a hardcoded value. We can refine it in the future if we'll have good estimates for other overhead sources.

File: src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/users/UserGenerator.kt [hunk 3]
```diff
@@ -4,6 +4,11 @@ import com.atlassian.performance.tools.jiraactions.api.memories.User
 import com.atlassian.performance.tools.virtualusers.api.VirtualUserOptions
 import net.jcip.annotations.ThreadSafe
 
+/**
+ * Supplies a user. The user can be created dynamically or it can find an existing user.
```
Comment (wyrzyk) [inline src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/users/UserGenerator.kt]: So `UserGenerator#generateUser` name is a little bit misleading. At least docs explain it now. 👍

Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/virtualusers/api/users/UserGenerator.kt]: True that. It should be called `UsersSource`, `UserSupplier` or something like that.

File: src/main/kotlin/com/atlassian/performance/tools/virtualusers/LoadTest.kt [hunk 4]
```diff
@@ -58,18 +56,31 @@ internal class LoadTest(
...
+        val users = (1..load.virtualUsers)
```
Comment (wyrzyk) [inline src/main/kotlin/com/atlassian/performance/tools/virtualusers/LoadTest.kt]: If `UserGenerator` is thread-safe, then we can generate it in parallel.

Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/virtualusers/LoadTest.kt]: All of them should be, we communicated that expectation via `@ThreadSafe`. Therefore we always do it in parallel.

File: src/main/kotlin/com/atlassian/performance/tools/virtualusers/LoadTest.kt [hunk 5]
```diff
@@ -58,18 +56,31 @@ internal class LoadTest(
...
+            .map { pool.submit(Callable { userGenerator.generateUser(options) }) }
+            .map { it.finishBy(deadline, logger) }
+        pool.shutdownNow()
```
Comment (wyrzyk) [inline src/main/kotlin/com/atlassian/performance/tools/virtualusers/LoadTest.kt]: What do you think about `shutdownNow ` in `finally` block?

Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/virtualusers/LoadTest.kt]: Haha 😀 you got me there
I'll `finally` do [it](https://bitbucket.org/atlassian/jira-performance-tests/pull-requests/76/jperf-123-a-sample-test-for-running-dc/diff#comment-78922842) 😉

Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/virtualusers/LoadTest.kt]: Raised and assigned https://ecosystem.atlassian.net/browse/JPERF-546

Comment (bogdanb) [reply src/main/kotlin/com/atlassian/performance/tools/virtualusers/LoadTest.kt]: I wonder if that’s why my assertions inside Actions were breaking tests.

Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/virtualusers/LoadTest.kt]: VU threads are fire-and-forget, so the unhappy-path-shutdown wouldn't matter. Any `Error` thrown from an `Action` will kill the VU and result in fewer measurements.

Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/virtualusers/LoadTest.kt]: I pushed in `AbruptExecutorService` usage

**Pull request merged**